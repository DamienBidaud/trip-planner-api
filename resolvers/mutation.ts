import { sign } from "jsonwebtoken";
import { hash, compare } from "bcryptjs";
import { Context } from "../index.d";

const signup = async (
  _: unknown,
  args: { email: string; password: string; username: string },
  context: Context
) => {
  const password = await hash(args.password, 10);
  const author = await context.prisma.author.create({
    data: {
      ...args,
      password,
    },
  });

  const token = sign(
    {
      authorId: author.id,
    },
    "DEV_ENV"
  );

  return {
    token,
    author,
  };
};

const login = async (_: unknown, args: { email: string; password: string}, context: Context) => {
  const author = await context.prisma.author.findOne(
    {
      where: {
        email: args.email,
      },
    }
  );

  if (!author) {
    throw new Error(`No author with the email ${args.email}.`);
  }

  const valid = await compare(args.password, author.password);

  if (!valid) {
    throw new Error(`The password is incorrect.`);
  }

  const token = sign(
    {
      authorId: author.id,
    },
    "DEV_ENV"
  );

  return {
    token,
    author,
  };
};

export default { signup, login };
