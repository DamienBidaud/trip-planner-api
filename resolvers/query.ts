import { PrismaClient, PrismaClientOptions, Author } from "@prisma/client";
import { verify } from "jsonwebtoken";
import { getActivity } from "../utils/activity";
type Context = {
  prisma: PrismaClient<PrismaClientOptions, never>;
};

const validateAuth = (_: any, args: { token: string }, context: Context) => {
  const { token } = args;
  try {
    const decoded = verify(token, "DEV_ENV") as {
      authorId: number;
      iat: number;
    };
    return decoded.authorId;
  } catch (err) {
    console.error(err);

    return null;
  }
};

const trip = async (
  _parent: unknown,
  args: { tripId: string },
  context: Context
) => {
  try {
    const allTrips = await context.prisma.trip.findOne({
      where: { id: parseInt(args.tripId, 10) },
    });
    return allTrips;
  } catch (error) {
    console.error(error);
    // throw error;
  }
};

const author = async (_: any, args: { token: string }, context: Context) => {
  const authorId = validateAuth(_, args, context);
  if (authorId) {
    return await context.prisma.author.findOne({
      where: {
        id: authorId,
      },
    });
  }

  return null;
};

const activity = (_: any, args: {activityID: string}, context: Context) =>
  getActivity(parseInt(args.activityID, 10), context);

const locations = (_: any, args: object, context: Context) => {
    return context.prisma.location.findMany();
};

export default {
  validateAuth,
  trip,
  author,
  activity,
  locations
};
