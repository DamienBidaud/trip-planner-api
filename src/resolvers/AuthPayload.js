const author = async (root, args, context, info) => {
  return await context.prisma.query.author({
    where: {
      id: root.author.id
    }
  }, info)
}

module.exports = {
  author
};