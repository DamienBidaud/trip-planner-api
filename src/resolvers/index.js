const Mutation = require('./Mutation')
const AuthPayload = require('./AuthPayload')
const Query = require("./query");
module.exports = {
  Query,
  Mutation,
  AuthPayload
}