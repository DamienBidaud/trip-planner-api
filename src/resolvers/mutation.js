const bcrypt = require("bcryptjs");
const jwt = require("jsonwebtoken");
const { findLocation, addLocation, getActivity, getTrip } = require("../utils");

const signup = async (root, args, context, info) => {
  const password = await bcrypt.hash(args.password, 10);
  const author = await context.prisma.mutation.createAuthor(
    {
      data: {
        ...args,
        password
      }
    },
    `{ id }`
  );

  const token = jwt.sign(
    {
      authorId: author.id
    },
    process.env.APP_SECRET
  );

  return {
    token,
    author
  };
};

const login = async (root, args, context, info) => {
  const author = await context.prisma.query.author(
    {
      where: {
        email: args.email
      }
    },
    `{ id password }`
  );

  if (!author) {
    throw new Error(`No author with the email ${args.email}.`);
  }

  const valid = await bcrypt.compare(args.password, author.password);

  if (!valid) {
    throw new Error(`The password is incorrect.`);
  }

  const token = jwt.sign(
    {
      authorId: author.id
    },
    process.env.APP_SECRET
  );

  return {
    token,
    author
  };
};

const createActivity = async (root, args, context, info) => {
  const { tripID, title, locations, price, startDate, endDate, type } = args;
  const trip = await context.prisma.query.trip(
    {
      where: {
        id: tripID
      }
    },
    `{id}`
  );

  if (!trip) {
    throw new Error(`No trip with id ${tripID}`);
  }

  const newActivity = await context.prisma.mutation.createActivity(
    {
      data: {
        title,
        locations,
        price,
        startDate,
        endDate,
        type
      }
    },
    `{ id }`
  );
  context.prisma.mutation.updateTrip({
    data: {
      activities: {
        connect: {
          id: newActivity.id
        }
      }
    },
    where: {
      id: tripID
    }
  });

  return getActivity(newActivity.id, context, info);
};

const createTrip = async (root, args, context, info) => {
  const { authorId, title } = args;
  const author = await context.prisma.query.author(
    {
      where: {
        id: authorId
      }
    },
    `{id}`
  );

  if (!author) {
    throw new Error(`No author with id ${authorId}`);
  }

  const newTrip = await context.prisma.mutation.createTrip(
    {
      data: {
        title
      }
    },
    `{ id }`
  );
  context.prisma.mutation.updateAuthor({
    data: {
      trips: {
        connect: {
          id: newTrip.id
        }
      }
    },
    where: {
      id: authorId
    }
  });

  return getTrip(newTrip.id, context, info);
};

const updateActivity = async (root, args, context, info) => {
  const {
    activityID,
    title,
    locations,
    price,
    startDate,
    endDate,
    type,
    description
  } = args;

  return context.prisma.mutation.updateActivity({
    data: {
      title,
      locations,
      price,
      startDate,
      endDate,
      type,
      description
    },
    where: {
      id: activityID
    }
  }, info)
};

const deleteActivity = async (root, args, context, info) => {
  return context.prisma.mutation.deleteActivity({
    where: {
      id: args.activityID
    }
  }, info)
};

const requestFriendship = async (root, args, context, info) => {
  const { authorId, authorRequestedId } = args;

  context.prisma.mutation.updateAuthor({
    data: {
      friends: {
        connect: {
          id: authorRequestedId
        }
      }
    },
    where: {
      id: authorId
    }
  });

  context.prisma.mutation.updateAuthor({
    data: {
      friendshipRequested: {
        connect: {
          id: authorId
        }
      }
    },
    where: {
      id: authorRequestedId
    }
  });
};

module.exports = {
  signup,
  login,
  createActivity,
  updateActivity,
  deleteActivity,
  createTrip,
  requestFriendship
};
