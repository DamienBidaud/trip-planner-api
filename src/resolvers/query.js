const jwt = require("jsonwebtoken");
const { getActivity, getTrip } = require("../utils");

const validateAuth = (_, args, context, info) => {
  const { token } = args;
  try {
    var decoded = jwt.verify(token, process.env.APP_SECRET);
    return decoded.authorId;
  } catch (err) {
    console.error(err);
    return null;
  }
};

const activity = (_, args, context, info) => getActivity(args.activityID, context, info);

const trip = (_, args, context, info) => getTrip(args.tripId, context, info)

const locations = (_, args, context, info) => {
  return context.prisma.query.locations({}, info);
}

const author = (_, args, context, info) => {
  const authorId = validateAuth(_, args, context, info);
  if(!!authorId) {
    return context.prisma.query.author(
      {
        where: {
          id: authorId
        }
      },
      info
    );
  }
  return null;
}

module.exports = {
  validateAuth,
  activity,
  locations,
  author,
  trip
};
