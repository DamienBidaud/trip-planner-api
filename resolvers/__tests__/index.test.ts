import resolvers from "../";

describe("resolvers", () => {
  it("Should match snapshot", () => {
    expect(resolvers).toMatchSnapshot();
  });
});
