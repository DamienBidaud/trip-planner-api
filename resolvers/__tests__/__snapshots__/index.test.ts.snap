// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolvers Should match snapshot 1`] = `
Object {
  "Mutation": Object {
    "login": [Function],
    "signup": [Function],
  },
  "Query": Object {
    "activity": [Function],
    "author": [Function],
    "locations": [Function],
    "trip": [Function],
    "validateAuth": [Function],
  },
}
`;
